retrieval/discovery/file_test.go

```diff
@@ -0,0 +1,93 @@
+package discovery
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+func TestFileSD(t *testing.T) {
+	testFileSD(t, ".yml")
+	testFileSD(t, ".json")
+	os.Remove("fixtures/_test.yml")
+	os.Remove("fixtures/_test.json")
+}
+
+func testFileSD(t *testing.T, ext string) {
+	// As interval refreshing is more of a fallback, we only want to test
+	// whether file watches work as expected.
+	fsd := NewFileDiscovery([]string{"fixtures/_*" + ext}, 1*time.Hour)
+
+	ch := make(chan *config.TargetGroup)
+	go fsd.Run(ch)
+	defer fsd.Stop()
+
+	select {
+	case <-time.After(15 * time.Millisecond):
+		// Expected.
+	case tg := <-ch:
+		t.Fatalf("Unexpected target group in file discovery: %s", tg)
+	}
+
+	newf, err := os.Create("fixtures/_test" + ext)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer newf.Close()
+
+	f, err := os.Open("fixtures/target_groups" + ext)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer f.Close()
+
+	_, err = io.Copy(newf, f)
+	if err != nil {
+		t.Fatal(err)
+	}
+	newf.Close()
+
+	// The files contain two target groups which are read and sent in order.
+	select {
+	case <-time.After(15 * time.Millisecond):
+		t.Fatalf("Expected new target group but got none")
+	case tg := <-ch:
+		if tg.String() != fmt.Sprintf("file:fixtures/_test%s:0", ext) {
+			t.Fatalf("Unexpected target group", tg)
+		}
+	}
+	select {
+	case <-time.After(15 * time.Millisecond):
+		t.Fatalf("Expected new target group but got none")
```

Comment:
I think in test cases where the *expected* case is the non-timer one (like here), it might be better to omit the timer case completely (and let the test simply time out) or set the timer much larger (like seconds). The latter way we still get a good error message, and it should almost never happen. Otherwise, who knows whether the tests could become flaky on slow systems like Travis, when there is a hickup...